File: app/src/main/java/com/riders/thelab/core/utils/LabAddressesUtils.java
package com.riders.thelab.core.utils;

import android.location.Address;
import android.location.Geocoder;
import android.location.Location;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class LabAddressesUtils {

    private LabAddressesUtils() {
    }

    public static Address getDeviceAddress(final Geocoder geocoder, final Location location) {
        Timber.i("getDeviceAddress");
        try {
            List<Address> addresses =
                    geocoder.getFromLocation(
                            location.getLatitude(),
                            location.getLongitude(),
                            1);
            Timber.e("addresses : %s", addresses);

            //get the address
            return addresses.get(0);
        } catch (IOException | NullPointerException e) {
            e.printStackTrace();
        }

        return null;
    }


    public static Single<List<Address>> getRXAddress(
            final Geocoder geoCoder,
            final double latitude,
            final double longitude
    ) {
        return new Single<List<Address>>() {

            @Override
            protected void subscribeActual(@NotNull SingleObserver<? super List<Address>> observer) {
                List<Address> addressList = new ArrayList<>();

                try {
                    addressList = geoCoder.getFromLocation(latitude, longitude, 1);

                } catch (Exception exception) {

                }
                if (addressList.isEmpty()) {
                    observer.onSuccess(addressList);
                } else {
                    observer.onError(new Throwable());
                }
            }

        }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static String buildAddress(final Address address) {
        String finalCity = "";
        StringBuilder addressStringBuilder = new StringBuilder();

        String street = address.getFeatureName() + ", " + address.getThoroughfare();
        String locality = address.getLocality();
        String postalCode = address.getPostalCode();
        String departmentName = address.getSubAdminArea();
        String regionName = address.getAdminArea();
        String countryName = address.getCountryName();

        addressStringBuilder
                .append(street).append(" - ")
                .append(locality).append(" - ")
                .append(postalCode).append(" - ")
                .append(departmentName).append(" - ")
                .append(regionName).append(" - ")
                .append(countryName);

        Timber.d(addressStringBuilder.toString());

        return locality;
    }

}

File: app/src/main/java/com/riders/thelab/core/utils/LabAddressesUtils.kt
package com.riders.thelab.core.utils

import android.location.Address
import android.location.Geocoder
import android.location.Location
import timber.log.Timber
import java.io.IOException

object LabAddressesUtils {

    fun getDeviceAddress(geocoder: Geocoder, location: Location): Address? {
        Timber.i("getDeviceAddress")
        try {
            val addresses = geocoder.getFromLocation(
                location.latitude,
                location.longitude,
                1
            )
            Timber.e("addresses : %s", addresses)

            //get the address
            return addresses?.get(0)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
        return null
    }


    suspend fun getRXAddress(
        geoCoder: Geocoder,
        latitude: Double,
        longitude: Double
    ): List<Address>? {
        var addressList: List<Address> = ArrayList()
        try {
            addressList = geoCoder.getFromLocation(latitude, longitude, 1)!!
        } catch (exception: Exception) {
            Timber.e(exception)
        }
        return if (addressList.isEmpty()) {
            null
        } else {
            addressList
        }
    }

    fun buildAddress(address: Address): String? {
        val addressStringBuilder = StringBuilder()
        val street = address.featureName + ", " + address.thoroughfare
        val locality = address.locality
        val postalCode = address.postalCode
        val departmentName = address.subAdminArea
        val regionName = address.adminArea
        val countryName = address.countryName
        addressStringBuilder
            .append(street).append(" - ")
            .append(locality).append(" - ")
            .append(postalCode).append(" - ")
            .append(departmentName).append(" - ")
            .append(regionName).append(" - ")
            .append(countryName)
        val finalAddress = addressStringBuilder.toString()
        Timber.d(finalAddress)
        return locality
    }
}
